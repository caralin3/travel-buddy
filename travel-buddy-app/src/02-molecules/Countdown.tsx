import moment from 'moment';
import React from 'react';

export interface CountdownProps {
  date: string;
}

export const Countdown: React.FC<CountdownProps> = ({ date }) => {
  const INTERVAL = 1000;
  let intervalId: NodeJS.Timer;

  React.useEffect(() => {
    intervalId = setInterval(getCountdown, INTERVAL);

    return () => {
      clearInterval(intervalId);
    };
  }, []);

  const [days, setDays] = React.useState(0);
  const [hours, setHours] = React.useState(0);
  const [minutes, setMinutes] = React.useState(0);
  const [seconds, setSeconds] = React.useState(0);

  function getCountdown() {
    const eventTime = moment(date);
    const now = moment();
    let duration = moment.duration(eventTime.diff(now));

    if (eventTime.isSameOrBefore(now)) {
      clearInterval(intervalId);
    } else {
      const durationDays = Math.floor(duration.asDays());
      duration.subtract(moment.duration(durationDays, 'days'));

      const durationHours = duration.hours();
      duration.subtract(moment.duration(durationHours, 'hours'));

      const durationMinutes = duration.minutes();
      duration.subtract(moment.duration(durationMinutes, 'minutes'));

      const durationSeconds = duration.seconds();
      duration.subtract(moment.duration(durationSeconds, 'seconds'));

      setDays(durationDays);
      setHours(durationHours);
      setMinutes(durationMinutes);
      setSeconds(durationSeconds);
    }
  }

  return (
    <div className="d-flex align-items-center justify-content-center">
      <div>
        <p className="h2 m-0">{days}</p>
        <p>days</p>
      </div>
      <div>
        <p className="h2 px-3">:</p>
      </div>
      <div>
        <p className="h2 m-0">{hours}</p>
        <p>hours</p>
      </div>
      <div>
        <p className="h2 px-3">:</p>
      </div>
      <div>
        <p className="h2 m-0">{minutes}</p>
        <p>minutes</p>
      </div>
      <div>
        <p className="h2 px-3">:</p>
      </div>
      <div>
        <p className="h2 m-0">{seconds}</p>
        <p>seconds</p>
      </div>
    </div>
  );
};
