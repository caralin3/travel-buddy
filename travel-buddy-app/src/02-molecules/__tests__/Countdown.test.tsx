import { fireEvent, render, screen } from '@testing-library/react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import moment from 'moment';
import React from 'react';
import { Countdown, CountdownProps } from '../Countdown';

const CountdownTestComponent: React.FC<CountdownProps> = ({ date }) => {
  const [days, setDays] = React.useState(0);
  React.useEffect(() => {
    const id = setInterval(() => {
      setDays(1);
    }, 0);

    return () => {
      clearInterval(id);
    };
  }, [date]);
  return (
    <>
      <Countdown date={date} />
      <button title="days" onClick={() => setDays(days + 1)}>
        Set Days
      </button>
    </>
  );
};

const dateMock = moment().add(1, 'day');

describe('Countdown', () => {
  it('should render correctly', () => {
    const wrapper = shallow(<Countdown date="2022-03-12" />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('should get countdown correctly', async () => {
    const setIntervalSpy = jest.spyOn(window, 'setInterval');
    render(<Countdown date={dateMock.format('YYYY-MM-DD')} />);
    expect(setIntervalSpy).toHaveBeenCalled();
  });
});
