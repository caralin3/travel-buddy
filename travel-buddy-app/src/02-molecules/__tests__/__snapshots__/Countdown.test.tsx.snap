// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Countdown should render correctly 1`] = `
<div
  className="d-flex align-items-center justify-content-center"
>
  <div>
    <p
      className="h2 m-0"
    >
      0
    </p>
    <p>
      days
    </p>
  </div>
  <div>
    <p
      className="h2 px-3"
    >
      :
    </p>
  </div>
  <div>
    <p
      className="h2 m-0"
    >
      0
    </p>
    <p>
      hours
    </p>
  </div>
  <div>
    <p
      className="h2 px-3"
    >
      :
    </p>
  </div>
  <div>
    <p
      className="h2 m-0"
    >
      0
    </p>
    <p>
      minutes
    </p>
  </div>
  <div>
    <p
      className="h2 px-3"
    >
      :
    </p>
  </div>
  <div>
    <p
      className="h2 m-0"
    >
      0
    </p>
    <p>
      seconds
    </p>
  </div>
</div>
`;
